src/tmpfiles/tmpfiles.c

```diff
@@ -1547,42 +1548,71 @@ static int create_directory_or_subvolume(Item *i, const char *path) {
                          * specifically set up a btrfs subvolume for the root
                          * dir too. */
 
-                        r = -ENOTTY;
+                        subvol = false;
                 else {
-                        RUN_WITH_UMASK((~i->mode) & 0777)
-                                r = btrfs_subvol_make(i->path);
+                        RUN_WITH_UMASK((~mode) & 0777)
+                                r = btrfs_subvol_make_fd(pfd, basename(path));
                 }
         } else
                 r = 0;
 
-        if (IN_SET(i->type, CREATE_DIRECTORY, TRUNCATE_DIRECTORY) || r == -ENOTTY)
+        if (!subvol || r == -ENOTTY)
                 RUN_WITH_UMASK(0000)
-                        r = mkdir_label(i->path, i->mode);
+                        r = mkdirat_label(pfd, basename(path), mode);
 
         if (r < 0) {
                 int k;
 
                 if (!IN_SET(r, -EEXIST, -EROFS))
-                        return log_error_errno(r, "Failed to create directory or subvolume \"%s\": %m", i->path);
+                        return log_error_errno(r, "Failed to create directory or subvolume \"%s\": %m", path);
 
-                k = is_dir(i->path, false);
+                k = is_dir(path, false);
```

Comment:
this should be is_path_at() or something like that, no?

Reply:
you meant is_dir_fd() ?

It should indeed but it didn't look unsafe if we keep operate on the path at a first glance but for completeness I'll fix it.